Regression 0.23.0 -> 0.23.1: `Cannot create directory, parent path is a file: b'C:'` during checkout

We noticed a regression between dulwich 0.23.0 and 0.23.1. The following snippet passes with `dulwich<0.23.1` but fails with `dulwich>=0.23.1` (only on Windows - it still passes on Linux):

<details>

<summary>pytest script</summary>

```
from dulwich import porcelain
from dulwich.repo import Repo


def test_dulwich_checkout(tmp_path) -> None:
    repo_path = tmp_path / 'repo'
    repo_path.mkdir()

    repo = Repo.init(str(repo_path))

    # first commit
    (repo_path / 'foo').write_text('foo', encoding='utf-8')
    repo.stage(['foo'])
    first_commit = repo.do_commit(
        committer=b'User <user@example.com>',
        author=b'User <user@example.com>',
        message=b'msg',
        no_verify=True,
    )

    # second commit in new subdirectory
    (repo_path / 'dir').mkdir()
    (repo_path / 'dir' / 'bar').write_text('bar', encoding='utf-8')
    repo.stage(['dir/bar'])

    second_commit = repo.do_commit(
        committer=b'User <user@example.com>',
        author=b'User <user@example.com>',
        message=b'msg',
        no_verify=True,
    )

    porcelain.checkout_branch(repo, first_commit)
    porcelain.checkout_branch(repo, second_commit)
```

</details>



This is the error:

<details>

<summary>error</summary>

```
tmp_path = WindowsPath('C:/Users/randy/AppData/Local/Temp/pytest-of-Randy/pytest-1382/test_dulwich_checkout0')

    def test_dulwich_checkout(tmp_path) -> None:
        repo_path = tmp_path / 'repo'
        repo_path.mkdir()

        repo = Repo.init(str(repo_path))

        # first commit
        (repo_path / 'foo').write_text('foo', encoding='utf-8')
        repo.stage(['foo'])
        first_commit = repo.do_commit(
            committer=b'User <user@example.com>',
            author=b'User <user@example.com>',
            message=b'msg',
            no_verify=True,
        )

        # second commit in new subdirectory
        (repo_path / 'dir').mkdir()
        (repo_path / 'dir' / 'bar').write_text('bar', encoding='utf-8')
        repo.stage(['dir/bar'])

        second_commit = repo.do_commit(
            committer=b'User <user@example.com>',
            author=b'User <user@example.com>',
            message=b'msg',
            no_verify=True,
        )

        porcelain.checkout_branch(repo, first_commit)
>       porcelain.checkout_branch(repo, second_commit)

tests\test_dulwich.py:34:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
C:\Users\randy\AppData\Local\pypoetry\Cache\virtualenvs\dt-kHwmRpoB-py3.12\Lib\site-packages\dulwich\porcelain.py:2750: in checkout_branch
    return checkout(repo, target, force=force)
C:\Users\randy\AppData\Local\pypoetry\Cache\virtualenvs\dt-kHwmRpoB-py3.12\Lib\site-packages\dulwich\porcelain.py:2684: in checkout
    update_working_tree(
C:\Users\randy\AppData\Local\pypoetry\Cache\virtualenvs\dt-kHwmRpoB-py3.12\Lib\site-packages\dulwich\index.py:1889: in update_working_tree
    _transition_to_file(
C:\Users\randy\AppData\Local\pypoetry\Cache\virtualenvs\dt-kHwmRpoB-py3.12\Lib\site-packages\dulwich\index.py:1697: in _transition_to_file
    _ensure_parent_dir_exists(full_path)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

full_path = b'C:\\Users\\randy\\AppData\\Local\\Temp\\pytest-of-Randy\\pytest-1382\\test_dulwich_checkout0\\repo\\dir\\bar'

    def _ensure_parent_dir_exists(full_path: bytes) -> None:
        """Ensure parent directory exists, checking no parent is a file."""
        parent_dir = os.path.dirname(full_path)
        if parent_dir and not os.path.exists(parent_dir):
            # Check if any parent in the path is a file
            parts = parent_dir.split(os_sep_bytes)
            for i in range(len(parts)):
                partial_path = os_sep_bytes.join(parts[: i + 1])
                if (
                    partial_path
                    and os.path.exists(partial_path)
                    and not os.path.isdir(partial_path)
                ):
                    # Parent path is a file, this is an error
>                   raise OSError(
                        f"Cannot create directory, parent path is a file: {partial_path!r}"
                    )
E                   OSError: Cannot create directory, parent path is a file: b'C:'
```

</details>


Is this a bug or do we use dulwich in a wrong way?

--- a/NEWS
+++ b/NEWS
@@ -12,6 +12,9 @@
  * Add colorized diff support for the ``show`` command with ``--color``
    argument. (Jelmer Vernooĳ, #1741)
 
+ * Fix Windows path handling in ``_ensure_parent_dir_exists`` to correctly
+   handle drive letters during checkout operations. (Jelmer Vernooĳ, #1751)
+
 0.24.1	2025-08-01
 
  * Require ``typing_extensions`` on Python 3.10.

--- a/dulwich/index.py
+++ b/dulwich/index.py
@@ -1595,19 +1595,31 @@ def _ensure_parent_dir_exists(full_path: bytes) -> None:
     """Ensure parent directory exists, checking no parent is a file."""
     parent_dir = os.path.dirname(full_path)
     if parent_dir and not os.path.exists(parent_dir):
-        # Check if any parent in the path is a file
-        parts = parent_dir.split(os_sep_bytes)
-        for i in range(len(parts)):
-            partial_path = os_sep_bytes.join(parts[: i + 1])
-            if (
-                partial_path
-                and os.path.exists(partial_path)
-                and not os.path.isdir(partial_path)
-            ):
-                # Parent path is a file, this is an error
+        # Walk up the directory tree to find the first existing parent
+        current = parent_dir
+        parents_to_check: list[bytes] = []
+
+        while current and not os.path.exists(current):
+            parents_to_check.insert(0, current)
+            new_parent = os.path.dirname(current)
+            if new_parent == current:
+                # Reached the root or can't go up further
+                break
+            current = new_parent
+
+        # Check if the existing parent (if any) is a directory
+        if current and os.path.exists(current) and not os.path.isdir(current):
+            raise OSError(
+                f"Cannot create directory, parent path is a file: {current!r}"
+            )
+
+        # Now check each parent we need to create isn't blocked by an existing file
+        for parent_path in parents_to_check:
+            if os.path.exists(parent_path) and not os.path.isdir(parent_path):
                 raise OSError(
-                    f"Cannot create directory, parent path is a file: {partial_path!r}"
+                    f"Cannot create directory, parent path is a file: {parent_path!r}"
                 )
+
         os.makedirs(parent_dir)
 
 

--- a/tests/test_index.py
+++ b/tests/test_index.py
@@ -2908,3 +2908,50 @@ class TestUpdateWorkingTree(TestCase):
 
         # file2 should still be a directory
         self.assertTrue(os.path.isdir(file2_path))
+
+    def test_ensure_parent_dir_exists_windows_drive(self):
+        """Test that _ensure_parent_dir_exists handles Windows drive letters correctly."""
+        from dulwich.index import _ensure_parent_dir_exists
+
+        # Create a temporary directory to work with
+        with tempfile.TemporaryDirectory() as tmpdir:
+            # Test normal case (creates directory)
+            test_path = os.path.join(tmpdir, "subdir", "file.txt").encode()
+            _ensure_parent_dir_exists(test_path)
+            self.assertTrue(os.path.exists(os.path.dirname(test_path)))
+
+            # Test when parent is a file (should raise error)
+            file_path = os.path.join(tmpdir, "testfile").encode()
+            with open(file_path, "wb") as f:
+                f.write(b"test")
+
+            invalid_path = os.path.join(
+                tmpdir.encode(), b"testfile", b"subdir", b"file.txt"
+            )
+            with self.assertRaisesRegex(
+                OSError, "Cannot create directory, parent path is a file"
+            ):
+                _ensure_parent_dir_exists(invalid_path)
+
+            # Test with nested subdirectories
+            nested_path = os.path.join(tmpdir, "a", "b", "c", "d", "file.txt").encode()
+            _ensure_parent_dir_exists(nested_path)
+            self.assertTrue(os.path.exists(os.path.dirname(nested_path)))
+
+            # Test that various path formats are handled correctly by os.path.dirname
+            # This includes Windows drive letters, UNC paths, etc.
+            # The key is that we're using os.path.dirname which handles these correctly
+            import platform
+
+            if platform.system() == "Windows":
+                # Test Windows-specific paths only on Windows
+                test_cases = [
+                    b"C:\\temp\\test\\file.txt",
+                    b"D:\\file.txt",
+                    b"\\\\server\\share\\folder\\file.txt",
+                ]
+                for path in test_cases:
+                    # Just verify os.path.dirname handles these without errors
+                    parent = os.path.dirname(path)
+                    # We're not creating these directories, just testing the logic doesn't fail
+                    self.assertIsInstance(parent, bytes)

